### ISSUE TITLE
Small picture (logo) per card 

### ISSUE THREAD

**Issue by x-jokay** [brarcher/loyalty-card-locker#86]
It would be great to be able to set a small picture per card on the overview.

Or even having an overview with the logo and name of the card only:
https://c.mobilegeeks.de/wp-content/uploads/2013/02/stocard-04.jpg

**Comment by brarcher** [comment]
Before creating this application I took a look at Stocard and a few others and how they worked. My concern with packaging the logos in the app is a legal one ([discussion here](https://www.quora.com/Can-I-use-another-companys-logo-on-my-website)). It is not clear to me if I would need to get permission from each company in writing prior to packaging their logo into the application and using it. Stocard's model appears to be based on partnerships with retailers. Although I do not know for sure, I suspect Stocard has the permission of the 1800 companies (as reported on Google Play) whose logo it uses.

It is unfortunately more complicated than all of that, however. This application is free (as in speech and beer) because of the license I have selected (GPLv3). To provide the assets with the same freedom if a logo from another company were to be included the company would need to give permission not only to me but to anyone who would fork this project, not knowing what they may do with it.

In the end, I am glad to provide an application that I (and hopefully others) find useful. I do not want to expose myself to legal trouble, nor be able to provide the application with the level of freedom and flexibility I expect. 

Sadly, although your feature request would be really nice to have, I must decline it. Hopefully the reason is understood.

**Comment by x-jokay** [comment]
The idea was not to package the logos directly into the application but to provide the possibility set a picture per card.

**Comment by brarcher** [comment]
Oh, that sounds do-able.

**Comment by brarcher** [comment]
How would that work, in your mind? Would the user take a picture with the camera of a logo? Would they import a photo? Something else?

**Comment by x-jokay** [comment]
For stocard I had taken some picture from a google image search and cropped it before setting it in the app.

You may link to an existing picture from the user's gallery or better import / copy it to some sub-dir of the loyality-card-locker data dir?

**Comment by x-jokay** [comment]
Or do directly a base64 encode of the picture.

**Comment by ni720** [comment]
That would significantly raise the acceptance in some shops for this app. As soon as an employee sees the logo of his company he is more likely to accept scanning the loyalty card from an app. Some shops however have a strict no-app-policy.

**Comment by brarcher** [comment]
> Some shops however have a strict no-app-policy

Really? I've not heard of that before. Interesting.

I am for the idea of having a user selectable logo displayed with the summary on the main page and with the barcode. What is holding back the idea is a nice layout design for when there is and is not a logo present. I'm not the best that that, and am hoping that someone with an idea may contribute a suggestion.

**Comment by x-jokay** [comment]
Why not using a default card icon as kind of placeholder if no picture was set?

**Comment by x-jokay** [comment]
![1_overview](https://cloud.githubusercontent.com/assets/18613935/23198401/9a141340-f8c7-11e6-92f6-962bf68b2aca.png)
![2_edit-nopic](https://cloud.githubusercontent.com/assets/18613935/23198403/9a45581a-f8c7-11e6-9dbd-aa9f23a9b86c.png)
![3_show-nopic](https://cloud.githubusercontent.com/assets/18613935/23198404/9a458d76-f8c7-11e6-8fca-e2345b605ba6.png)
![4_show-nopic-adv](https://cloud.githubusercontent.com/assets/18613935/23198405/9a46af08-f8c7-11e6-9df4-defb486b7d0b.png)
![5_edit-pic](https://cloud.githubusercontent.com/assets/18613935/23198406/9a47b240-f8c7-11e6-9122-d6acb10173dd.png)
![6_show-pic](https://cloud.githubusercontent.com/assets/18613935/23198407/9a4b0fda-f8c7-11e6-92ce-921851df3946.png)
![7_show-pic-menu](https://cloud.githubusercontent.com/assets/18613935/23198402/9a2ec79e-f8c7-11e6-8491-a9194de71f77.png)

**Comment by x-jokay** [comment]
Some samples without and with icon.

**Comment by x-jokay** [comment]
I'm not sure if it's really necessary to have the barcode displayed on the main page?

**Comment by jimmyswimmy** [comment]
This last set of pictures offers an interesting idea - it would be awesome to include pictures of the card.  I was shopping at Costco and the cashier wanted to see the card, but then scanned my phone... but he said if I'd had pictures he would have accepted that.  I'd be interested to try to implement this (know nothing about android programming but I'm a quick study) if x-jokay has all of those icons etc.  I hate front end work.

**Comment by x-jokay** [comment]
The screenshots were taken from the app stocard, I don't "own" those icons.

**Comment by brarcher** [comment]
If the issue is icons, I've used a number of icons from [The Noun Project](thenounproject.com) for other applications. @jimmyswimmy, if you are interested in taking this on and need some icons, maybe give that a look. The only thing which would need to be done is to ensure that the icons are release under a Creative Commons license and give credit in the [README](https://github.com/brarcher/loyalty-card-locker/blob/master/README.md) and in the [About dialog](https://github.com/brarcher/loyalty-card-locker/blob/master/app/src/main/java/protect/card_locker/MainActivity.java#L164). [Here](https://github.com/brarcher/loyalty-card-locker/blob/master/app/src/main/java/protect/card_locker/MainActivity.java#L164) is an example of adding references for image resources.

**Comment by ioangogo** [comment]
Has any progress happened in this issue yet, I am tempted to work on this issue as I also feel like it would be useful in shops that might be a bit hesitant on scanning a barcode from this app

I'm currently thinking about how display it on the card page, I'll upload some scribbles at a later point as I am currently not near a laptop/desktop

**Comment by brarcher** [comment]
I've not made any progress on this. I encourage you to give it a shot!

**Comment by x-jokay** [comment]
Not as I'm aware of :)

Anyway, mobile payment solution providers start to integrate customer cards directly and it will be used automatically during payment process.

We may continue to use this app for customer cards which are not yet supported.

**Comment by ioangogo** [comment]
Currently i have 3 ideas, 2 i have mocked up in GIMP and the other is just having the logo over the barcode. I like it being by the store name, but that only works with stores that have a logo like the national trust, or the body-shop, most have their name in the logo so having it above the barcode may work but does not allow for low resolution images.

Mockups:

1
![left](https://user-images.githubusercontent.com/7449028/28671759-b7f07394-72d5-11e7-9a92-40c39edc5395.png)
2
![right](https://user-images.githubusercontent.com/7449028/28671760-b7f0d744-72d5-11e7-8ffa-01c35c93bc05.png)

**Comment by ioangogo** [comment]
@x-jokay yeah, android pay supports quite a few now, although I did find this useful when visiting a national trust property and my farther had forgotten his card

**Comment by brarcher** [comment]
> but that only works with stores that have a logo like the national trust

I would expect that if the store had or did not have the name in the logo it would still be recognizable to the end user, and may not require having the store name and logo both.

Could you create a mock-up with the logo above the barcode, to see what that is like? 

> but does not allow for low resolution images

Do you mean that if one were to use a low resolution image it would not look good, so for a good experience one would need to source higher resolution images?

**Comment by ioangogo** [comment]
@brarcher on the resultion thing, yes I did mean that

I'll try to make a mock up of the logo above the barcode when I can.

**Comment by brarcher** [comment]
Thanks for taking this on, @ioangogo!

**Comment by ioangogo** [comment]
Here is a mock up of it above the barcode

![screenshot 31 jul 2017 5-34-51 pm](https://user-images.githubusercontent.com/7449028/28789324-d30081d4-761c-11e7-8061-735e47cf561e.png)

**Comment by brarcher** [comment]
I like the full sized barcode more, what do you think?

Another thing that I am thinking is perhaps the order of the information could be better, to put the more relevant information closer to the top. For example:

![image](https://user-images.githubusercontent.com/5264535/28789685-7d7116ca-75f4-11e7-9e77-8b71a02d5ead.png)

Or, another option might be to omit the store name if there is an icon:

![image](https://user-images.githubusercontent.com/5264535/28789762-bd9e735a-75f4-11e7-8631-e15e0352715b.png)

Do you have an opinion on the order of the presented data? Perhaps there is a better idea.

**Comment by ioangogo** [comment]
I personaly prefer the current set up, It just kind of flows with the large images at the bottom and the small text at the top

**Comment by ioangogo** [comment]
> Or do directly a base64 encode of the picture.
@x-jokay the app uses a sqlite database so it may be possible to use the SQLITE BLOB data type, but to keep the CSV exporting simple I might use your suggestion

**Comment by matthijskooijman** [comment]
Just to double-check: Is this issue about just adding a logo, or are photographs of the card itself also considered? It has been mentioned in the comments, but I'm not sure if they're also considered (I do think images of the card, front and back, really add something. Using the front side of the card as the "logo" in the overview of cards works really well in [mobile pocket](https://play.google.com/store/apps/details?id=at.bluesource.mobilepocket&hl=en) which I was using before). If it helps, I'm happy to create a separate issue for that, though I suspect these could be tied together, also in terms of implementation.

**Comment by x-jokay** [comment]
It was about adding a logo to the card (manually from the gallery) :)

I think a separate issue for this will be a good idea.

This issue can then be linked if it really makes sense for the implementation.

**Comment by ioangogo** [comment]
Sadly due to workload related reasons I will probably not be able to do this now, I hope someone else picks this up. I had written some code, but it was mostly updating the database, sadly the laptop it was on died and I haddent committed it to my fork

**Comment by Flavien06** [comment]
In the [andOTP ](https://github.com/andOTP/andOTP) project, a small logo can be added to the list.
Both applications have a fairly similar interface, it could be an integration idea, with a logo on the card list.
https://raw.githubusercontent.com/flocke/andOTP/master/assets/screenshots/main_activity.png

**Comment by x-jokay** [comment]
This looks really good.

May be we could get at least the small logos on the overview instead of the full card view itself?

**Comment by brarcher** [comment]
I agree, the layout used by andOTP looks pretty nice. That is doable.

The remaining questions I have are the following:

1. Where does the image come from?
1.1: If it is imported from the gallery, where does the user get the original image? Manually search for it online somehow?
1.2: If the image is a screenshot of the card itself, will the user expect to edit the photo to narrow in on the logo?
2. Will the logo images be backed up using the backup/restore feature?
2.1: As the backup feature currently only uses CSV files, does this mean a ZIP option will need to added which supports also backing up the images?

**Comment by x-jokay** [comment]
I see, we will never get logos for this app.

See old comments for questions.

Locally stored images, added by user manually in the app for each card, base64 encoded for saving and backup may be an option.

**Comment by gilcot** [comment]
According to me, the image should come either from the user's own gallery (no online search) or camera. You know, like the contact app does (and some other too.) 

For the second point, no zip should be needed: a directory with images named with their ID should be exported near the CSV files. But another good option may be to base64 encode them in the CSV...

**Comment by brarcher** [comment]
> I see, we will never get logos for this app.

@x-jokay I think you mean well. Just note that this app is a hobby for me and I do it for fun. There are also other projects that I like working on, so this app is not always my top priority. My motivation is to have open source tools which are useful to me. That others also find them valuable is great. Sometimes people request features for my apps, and that is good because it shows that people care. I'm not always motivated to add them, though.

From the information I have from the discussion, I think I have enough to take a try at it. The first attempt will likely be displaying the logo in the list view and the barcode view, and take whatever image the user provides from the gallery and stick it in the image view without an option to modify or resize it. Backups would not be supported at first, as I'll need to think about it more. For example, if a user adds a 1MB image, I do not want to base64 that and put it into a CSV file.

**Comment by x-jokay** [comment]
> From the information I have from the discussion, I think I have enough to take a try at it.

@brarcher Good to have you motivated and thank you :)

**Comment by brarcher** [comment]
After looking at how andOTP does its layout, this is the example layout I came up with for the card list:

![screenshot_1517100246](https://user-images.githubusercontent.com/5264535/35477826-ce3851de-039a-11e8-9c68-f94ae3ac88da.png)

In addition I also adapted from andOTP the default thumbnail, which is the first letter of the store name and a pastel background. The color is chosen by hashing the store name, so the color used is essentially random.

I like this layout much better, as it emphasizes the store name. Thoughts so far?

**Comment by x-jokay** [comment]
This looks really good.

**Comment by brarcher** [comment]
I'll make this change. The sad part is that the card list layout looking this good makes the layout with the barcode look pretty bad in comparison. (:

I'm looking into how stocard does their card layout with the area near the top with either the logo or a solid color. 

![image](https://user-images.githubusercontent.com/5264535/35483453-256318ac-0410-11e8-8d30-0a455dea741f.png)

If anyone knows an open source project that does something similar I can use as an example, let me know. Otherwise, I'll try to figure it out.

**Comment by brarcher** [comment]
How does the following look an an example of viewing a card? This is the "Pharmacy" card from the previous picture.

![screenshot_1517374845](https://user-images.githubusercontent.com/5264535/35605910-f47fb8ec-0619-11e8-8484-ed664925e1e3.png)

**Comment by brarcher** [comment]
@x-jokay Would you mind trying out this change and give me any feedback you may have?

[app-release.apk.zip](https://github.com/brarcher/loyalty-card-locker/files/1681834/app-release.apk.zip)

^ This does not have configurable images, as I want to get the layout and behavior working first. There are still some things to fix, but I'm curious of your opinion thus far.

**Comment by Flavien06** [comment]
That's pretty good.
Comments appear in very big.
![screenshot_20180131-145829](https://user-images.githubusercontent.com/1915755/35645492-498d40a4-06cc-11e8-8933-d7b5570fc5b9.png)

Can be done in 3 lines: Title (large bold), commentary (bold) and number.
![screenshot_20180131-1458229](https://user-images.githubusercontent.com/1915755/35645746-26d2eebe-06cd-11e8-8269-f8ceeab86b18.jpg)

**Comment by x-jokay** [comment]
@brarcher installed the apk and it looks very good, great job!

@Flavien06 it's currently the same behavior as for the release version. I do not enter notes for the cards anyway but it may be useful for someone :)

I would prefer the logo and the title only for the overview but it's only one opinion of many :)

**Comment by brarcher** [comment]
In my usage of the app I do agree that the number on the first screen has dubious value. I'll change the layout to omit the number on the first screen.

As for the note, it presently will not appear on the single card screen. I'll put the note on the second line in a smaller font. If later it turns out to be more valuable to have it on the single card view, maybe it can be removed from the card list screen.

**Comment by brarcher** [comment]
Thanks for the feedback. This is what I settled on for now:
![image](https://user-images.githubusercontent.com/5264535/35660305-458572a8-06d9-11e8-90b3-8ce197540bae.png)
![image](https://user-images.githubusercontent.com/5264535/35660310-4c7d45cc-06d9-11e8-9dc8-728d7f8f9745.png)
The edit screen is not yet changed. I'll need to think about how that should look. For now, though, most of the interaction with the app will be viewing cards, that that will have a nicer layout.

I'll create a release, but I'll mark it as not production ready. It will be released to the Play store, but on the Alpha or Beta channel for now. If that goes well and I do not get negative feedback I'll release it to the main channel.

**Comment by Flavien06** [comment]
Great
How to access the beta?

**Comment by brarcher** [comment]
I'll cut a release on GitHub later today. From there one can download the APK directly.

F-Droid will probably pick it up like a normal release, so in a few days one can download the APK from F-Droid signed with their key.

I'll put the release on the Play Store as well under the Beta stream. I do not know how one signs up for Beta streams on the Play Store.

**Comment by brarcher** [comment]
The release has been made and the APK has been uploaded to the Play Store on the Beta channel. To sign up for the Beta Channel, go to the following link:
https://play.google.com/apps/testing/protect.card_locker

Feel free to post any feedback on this thread or email me.

If I do not hear any negative feedback in the next 2 weeks I'll assume all is well and release it to the normal track on the Play Store.

**Comment by brarcher** [comment]
I've only heard positive feedback thus far. The beta version has been pushed to production on the Play store. If there are any issues with the new version either respond in this issue (if it is layout related) or open a new issue (if it is functional related).

The next step for the layout improvements is to modify the card editing to the new layout. I'll wait to see if there is any feedback from Play users and move forward from there. After that, if all goes well, changing the capital letter to a custom logo can be worked on.

**Comment by brarcher** [comment]
I've received additional feedback on the layout in issue https://github.com/brarcher/loyalty-card-locker/issues/194 from @denilsonsa. Namely,

- the number text was too long and was truncated sometimes, 
- the name of the card does not appear in the single card view but the name is valuable, and
- the note of the card does not appear but is valuable

The first item is a good point. It is easy enough to reduce the font size to accommodate more characters. I would like to have the card ID one one line, but understand that a card ID can be any length. If the font were reduced it could acccomidate maybe 20 characters on my example device, from:
![image](https://user-images.githubusercontent.com/5264535/35954940-833393ba-0c5a-11e8-8d6d-857dde358021.png)
to this:
![image](https://user-images.githubusercontent.com/5264535/35954944-8bd9771e-0c5a-11e8-9684-5f257b4428a0.png)
Is it preferable to allow the card ID to fill as many lines as it may take up? That is, something like:
![image](https://user-images.githubusercontent.com/5264535/35955587-3ca0f5c4-0c5e-11e8-8cb5-aa162132e15c.png)


For the third point, I'm not opposed to displaying the note if one has been entered. Perhaps something as simple as:
![image](https://user-images.githubusercontent.com/5264535/35955499-b19161da-0c5d-11e8-9c49-d12acb883b29.png)

About the second point about the store name, presently the only option for the area at the top is to display the first character of the name. The goal for the future is to display the logo. If the logo is displayed, the store name is unnecessary. Should there be middle ground when no logo is configured? Namely, should the name be listed under the color but above the barcode, or within the color area:

![image](https://user-images.githubusercontent.com/5264535/35955932-91e80f84-0c60-11e8-96ed-ad8dd4d137c1.png)

**Comment by denilsonsa** [comment]
**TL;DR:** Please always show both the note and the barcode content, and make sure they are not truncated (i.e. wrap them as needed).

Hi! I'm gonna present some real-world use-cases.

## About numbers and notes

There are some cases where the actual number in the barcode and the number printed on the card differ. For instance, I have a card that has `1234 5678 9012 3` printed, but the barcode actually reads `6064361234567890123` (i.e. the barcode includes a prefix). For this case, I store the human-readable number as a note.

Sometimes, the printed number is nicely formatted (`01-234.567.8901`), so I'd like to store this formatting somewhere, because it is easier to recognize as a valid number and easier to type manually. I'm currently storing it as a note.

In addition, some scanners (usually the laser ones) cannot read from the phone screen, or we are typing the number on a website, which means we really need the full number visible. Sometimes the barcode number is okay, other times the note is better.

There is one case where I store a "pin" code as a note (I know, not secure, I do not recommend doing that, but for this use-case it is fine). So, for this case I need both the number and the note visible.

## Summary

Please show both the barcode number and the note.

And please make sure they all fit into the screen.

Fitting into screen is tricky, because of different device sizes and device resolutions, and arbitrary barcode content length. I have a few ideas regarding that:

* The simplest implementation is to have a fixed font size and allow the text to wrap.
* A more complex implementation could be to auto-adjust the font-size within a certain range to scale the text as needed. We could aim for largest font size without wrapping. And if it still doesn't fit, just wrap.
* As a special case, it could try to wrap the text exactly in half, if it is longer than one line but fits in two lines. Still, I'm not sure this is worth the effort.

I'd go for the simple solution.

**Comment by denilsonsa** [comment]
**TL;DR:** Just read the **bold** text.

## About pictures

I like the idea of having a logo or a picture attached to each card. That's nice. However, be aware that adding such feature brings a lot of extra work.

* **How to import the picture**? Only from gallery or also from the camera? (I suppose Android already has an intent that solves this issue for us.)
* Can I try **grabbing a picture from the Internet** and use it? [Google Docs](https://docs.google.com/) has this feature, and it will likely be very useful for letting users find logos for their cards.
* After the picture is imported, the user might like to **edit it** (to fix distortion and rotation, crop the photo, tweak the colors/lighting/brightness/contrast). Sure, one could argue this is a job for a different app, but having this built-in can improve UX. And, yes, I know, this is a lot of effort to implement; as a developer I would like not to implement that. But as a user I would like to have that.
* After the picture is imported, **how to display it**? Because it will likely have a different aspect ratio than the widget. Should it be scaled to fit the area (no cropping), or scaled to cover the area (cropping). Should small images be scaled up? What if the user would like to view the picture as "full screen", including zoom?
* Cool! Now we have one picture imported into the card! Can I **replace the picture** with a new one? Surely, right?
* Can I **add a second picture**? As previously commented in this thread, people would like to keep both the front and the back of the card.
* Now that we have two pictures, **how to display them**? Sliding to the sides, like a carousel?
* Should we restrict to **only two** pictures? What if the user needs more?
* Okay, finally all of these questions have been solved! We are done, right? Well, not yet. People can export the cards as CSV, and pictures cannot go into CSV. How can we **export the data** from the app in a way that includes pictures? (Maybe a zip file with CSV and images, and the CSV includes a column with the picture names for each card.)
* Now that we can export everything, the app must be able to **import** it.

Do I want to discourage anyone from implementing this feature? ~~A little bit~~ Not at all! I just want to show the amount of work needed. We need to break it down by steps, by small tasks, and keep track of which tasks are still unfinished.

## About names, logos, colors

Once we have a picture or logo, should it replace the name? Yeah, I think so. I don't see much reason to keep both the name and the picture visible in the "view card" screen. If the picture is displayed in a carousel widget, we could have one "slide" with just the name and another "slide" with just the picture. That way, if someone really needs, they can just swipe to one of the sides and see the name.

Now I have another feature request: **please allow the user to pick their own colors for each card.** Why? Because each company has their own brand color, so it is easier to find the card if the colors match. Even employees might "trust" the app a bit more if the colors match.

But, if we are going to support custom colors, we must support **both background and foreground colors**.

## The end

Thank you for reading through my long wall of text. :)

**Comment by brarcher** [comment]
I appreciate the feedback about your use case, @denilsonsa. Thanks for the details.

> Please show both the barcode number and the note.

Having used the app more since the layout change, I agree. How does the following look? I've changed the text to auto-size within a certain range. The card ID will go until it runs out of screen. If there is a note, it will be the last line of the screen. In addition, the single letter "logo" is replaced with the store name. Here are some examples:

![image](https://user-images.githubusercontent.com/5264535/36059015-c52857d0-0dfc-11e8-8bde-0cc5c053128f.png)
![image](https://user-images.githubusercontent.com/5264535/36059024-1531f72c-0dfd-11e8-8c4c-518d3d73cfec.png)
![image](https://user-images.githubusercontent.com/5264535/36059025-1b1cf934-0dfd-11e8-844c-abdaffb487c3.png)
![image](https://user-images.githubusercontent.com/5264535/36059029-3a4a51d0-0dfd-11e8-9cf0-276aa7522d07.png)

> Do I want to discourage anyone from implementing this feature? ~A little bit~ Not at all! I just want to show the amount of work needed.

This is one reason I've been putting it off. I've another app (which keep track of budgets) that allows photos of receipts. Managing the photos, allowing them to be imported/exported, etc, is a bit of work. In that case the images have no edit option, but to use the images in this app something would need to be done. If someone takes a photo it will likely not look that nice. Letting the app fetch an images from the internet requires internet permission, and I want to avoid that. 
 
Getting colors and a better design for the main layout was an easier first step. (:

> please allow the user to pick their own colors for each card.

I'm thinking that after getting some of the layout issues resolved this will be the next step. Right now the color is chosen for you and is basically arbitrary.

**Comment by denilsonsa** [comment]
I like the new screenshots.

A couple comments/questions:

* What if the note does not fit one line?
* If, like I explained, I have a human-readable version of the number in the note, shouldn't the note have a font as big as the main number?
* As a bonus, can the user select (and copy) the displayed number and note? Which is something from #90. ;)

> Getting colors and a better design for the main layout was an easier first step. (:

Fully agree! And I agree that picking up colors seem like the next step.

**Comment by denilsonsa** [comment]
Just an idea I wanted to share…

In case we would like to let the user fetch logos from companies from the Internet, we could try letting the user navigate through this:

* https://commons.wikimedia.org/wiki/Category:Logos_of_companies_by_country
* https://commons.wikimedia.org/wiki/Category:Logos_of_organizations_by_country

On one side, the app would not ship with any logo, but just provide a browsing tool to find one. And it would get one from wikimedia commons.

However, I don't know the legal implications of that (hopefully none); and indeed it would require Internet access permission.

**Comment by brarcher** [comment]
> What if the note does not fit one line?

If the card id is too long it can overflow the screen; as the length of the id is unbounded this is possible. 

My prototype has the note field taking precedence over the id field. That is, if the card id is too long the note field will still be shown. Right now the note field is not limited in the number of lines. If this were so and the note were really big, the card Id could never appear.

![image](https://user-images.githubusercontent.com/5264535/36063991-5334bcee-0e52-11e8-8d53-1fdd4a7b2ce8.png)
![image](https://user-images.githubusercontent.com/5264535/36063992-5a0896a8-0e52-11e8-97a2-6e5042cbaca1.png)

> If, like I explained, I have a human-readable version of the number in the note, shouldn't the note have a font as big as the main number?

I've not figured out how to get that to work just yet. Using a text view that has its size adaptable, I find that if the card ID has large text it squishes the note field. What could be done for now is to increase the minimum font size for the note field:

![image](https://user-images.githubusercontent.com/5264535/36064120-298f0924-0e54-11e8-9dbb-54c629909a1c.png)

> As a bonus, can the user select (and copy) the displayed number and note?

Can do.

![image](https://user-images.githubusercontent.com/5264535/36064149-7da6b5de-0e54-11e8-86a0-d1723e5fa938.png)

> In case we would like to let the user fetch logos from companies from the Internet

I'd like to avoid giving this app Internet permission. This is for privacy reasons, although it does make the use case of adding an icon more difficult. It is possible to launch a browser from an intent, but I am not aware of doing so and getting a result back (e.g. an image file). It could be a two step process, namely (1) tell user to download icon from a webpage and launch browser, then (2) tell user to find the icon they downloaded and launch a file browser. It would be a little clunky, but would avoid the Internet permission.

> However, I don't know the legal implications of that (hopefully none)

I'm not a lawyer, but using the logos would be for "personal use", and it would be difficult to come after someone for downloading an image available from Wikimedia for use on their own phone.

Now, that list from wikimedia is interesting. Looking through some details, they source the images from elsewhere and post them as public domain (from what I saw). Trademark laws do apply, however. Here is an example statement from Apple about their logo:

> Use of the keyboard Apple Logo (Option-Shift-K) for commercial purposes without the prior written consent of Apple may constitute trademark infringement and unfair competition in violation of federal and state laws. Use of Apple trademarks may be prohibited, unless expressly authorized.

This app is not "commercial" (as I get no revenue from it). However, the last sentence attempts to deny all right except those explicitly granted. That said, [here](https://commons.wikimedia.org/w/index.php?search=apple+logo&title=Special:Search&go=Go&searchToken=ahdnom5rxiipblrxinezt18al#/media/File:Vectorized_Apple_gray_logo.svg) is the Apple logo on Wikimedia.

It would be feasible (technically, not sure legally yet) to have the logos embedded into the app. However, it would increase the size of the app significantly. For example, the stocard app has 3944 logo embedded in it. Compressed that is 9MB. There is also a database which maps the logos to countries.

Coming back to the issues about the store name and note, are the images I shared above inline with your use case and OK for now?

**Comment by denilsonsa** [comment]
Yes, they look great! Looks like a great improvement from the first iteration (0.19).

My comment about very very long notes and numbers comes from my developer background trying to think and handle corner cases and trying to abuse the software to see if/when it breaks. It does not come from any real use case (but maybe someone somewhere will try it).

I dislike the idea of embedding hundreds of images into the APK. That's a lot of wasted storage for most users, something very valuable for low-end phones. And there will be several requests of adding new images and updating existing ones.

**Comment by brarcher** [comment]
I'll push the changes from my screenshots and create a new release.

The next step will likely be the custom colors. Can you clarify what you mean by:

> But, if we are going to support custom colors, we must support both background and foreground colors.

Namely, what is the background and what is the foreground? In the following image, is the foreground blue and the background white? Does that refer to the top section with the store name?

![image](https://user-images.githubusercontent.com/5264535/36065432-c28c6950-0e68-11e8-9314-a18d56b5aa37.png)

**Comment by denilsonsa** [comment]
> Can you clarify what you mean ny […]

Indeed, in that case the background is blue and the text (foreground) is white.

But there are cases where the background will be light and the text must be dark (or some color). BestBuy would have a yellow background with black text; IKEA would have yellow on blue, or blue on yellow.

For instance, Gmail labels can be configured to a pair of colors (background/foreground).

**Comment by brarcher** [comment]
I've a prototype for selecting the colors of the card info header. The configurations look like the following:
![image](https://user-images.githubusercontent.com/5264535/36118211-4264e542-100a-11e8-8160-363acc2e0d9f.png)
where one can either pick from a small selection of colors:
![image](https://user-images.githubusercontent.com/5264535/36118406-ea4a632c-100a-11e8-9649-3c9400e4893a.png)
or pick or enter a custom color:
![image](https://user-images.githubusercontent.com/5264535/36118417-f7500e0a-100a-11e8-86d7-5c92a7405daf.png)

A default background color is selected at random (or for pre-existing cards it uses the same color used now). The color choices affect both the single card view as well as the main listing:

![image](https://user-images.githubusercontent.com/5264535/36119440-f4f48d2c-100d-11e8-8ac6-af483d85563e.png)

![image](https://user-images.githubusercontent.com/5264535/36119427-ef8ec28a-100d-11e8-8d4d-5ef6e656ea8c.png)


Anyone interested in giving it a try and providing feedback? The APK is:
[loyalty-card-keychain-v0.21b.apk.zip](https://github.com/brarcher/loyalty-card-locker/files/1717668/app-release.apk.zip)

Note that this does add entries to the database, so take a backup before installing it in case you want to restore to the previous release.

**Comment by devurandom** [comment]
Would it be possible to allow the user to use a logo instead of the coloured-single-letter in the list of cards? That would make it easier / quicker to find the desired card in the list.

**Comment by brarcher** [comment]
I do not see why it could not. Right now I'm satisfied with the app layout. If someone where interested in working on the logo feature I'd be happy to review the pull request.

**Comment by OmlineEditor** [comment]
> * **How to import the picture**?
to start, use only the download option from the gallery


> * Can I try **grabbing a picture from the Internet** and use it?
do not use the internet at all. no internet


> * After the picture is imported, the user might like to **edit it**
at the beginning make only a crop. just select the desired piece from the picture and that's it. just the right aspect ratio. it is only at the beginning then in the future we will add new functionality

> * After the picture is imported, **how to display it**?
![35660305-458572a8-06d9-11e8-90b3-8ce197540bae](https://user-images.githubusercontent.com/13348933/71787770-c9c44980-302c-11ea-8397-71cce109fefd.png)



> * People can export the cards as CSV, and pictures cannot go into CSV.
use base64 to import inside csv. now tah I have stored photos of contacts from the phone book. they are converted to base64


All problems can be solved. need to start acting. The idea was proposed already 2 years ago and there is no implementation yet. I hope this comes soon.

**Comment by OmlineEditor** [comment]
> I do not see why it could not. Right now I'm satisfied with the app layout. If someone where interested in working on the logo feature I'd be happy to review the pull request.

Today, exactly 2 years have passed since the opening of the ticket. 

you may not need to make a complex project to select a logo image.

you can make a simple logo for cards in the form of a picture of a certain size no more than 200x125 pixels (this is the ratio of the card 86x54 mm) the user can only upload small pictures that will be shown in the app. no cropping, no zooming, only simple images from the gallery no more than the specified size.

can you do that?

**Comment by TheLastProject** [comment]
@brarcher did say fairly clearly that they have no intent to work on this but will happily review a pull request introducing this feature. Sadly, to this day, nobody has created a pull request for this.

I may (no promises!) get to this at some point as I've been tackling a lot of little things in this project in my spare time, but if you really want this feature, maybe you should contribute some of your spare time implementing it. If you look at the pull requests section on GitHub, the vast majority get approved given the code is of acceptable quality and unit tests are written for them.

**Comment by OmlineEditor** [comment]
> maybe you should contribute some of your spare time implementing it.

sorry I don't know this programming language. I can't program for Android.
I will wait.

**Comment by x-jokay** [comment]
Patience is a virtue.

**Comment by brarcher** [comment]
@OmlineEditor, thanks for your interest in the project. The project is a hobby that I do for free. These days there are other things taking my free time, so I'm not able to contribute as much as before. This is the reason I've really only accepted pull requests from other contributors rather than creating new features myself. 

I'm happy that there are developers who are interested in the project and have submitted improvements or features. Not everyone is a developer or familiar with Android, so it is perfectly fine to report bugs, and suggest improvements and features and hope that someone picks them up. As the project is entirely run by volunteers, there is no expectation that anything in particular will be worked on. Contributions are based on what someone finds interesting. (:

My initial hesitance with this feature is the project possibly controlling the store logo assets. Expecting a user to go through the steps to download the logo and import it into the app removes this concern. All the other aspects of the feature are solvable from there, as @OmlineEditor's summary shows. Implementing the change will involve a bit of work -- if someone were interesting in giving it a try I'd help review the changes. 

> sorry I don't know this programming language. I can't program for Android.

That is fine. (:

Truth be told, this was one of the first Android apps I ever wrote. (And that is why the app is far from perfect -- there is a bit I would do differently if I had to do it over). If you were interested in trying to start with Android you are welcome to use contributing to this project as a learning opportunity, @OmlineEditor.
